src/main/scala/com/microsoft/hyperspace/index/rules/HPFilterIndexRule.scala

```diff
@@ -0,0 +1,166 @@
+          ResolverUtils.resolve(spark, ddColumns.indexed.head, filterColumnNames).isDefined
+        }(plan, index) &&
+        withReasonTag(
+          "Index does not contain required columns. Required columns: " +
+            s"[${filterColumnNames ++ projectColumnNames}], indexed & included columns: " +
+            s"[${ddColumns.indexed ++ ddColumns.included}") {
+          ResolverUtils
+            .resolve(
+              spark,
+              filterColumnNames ++ projectColumnNames,
+              ddColumns.indexed ++ ddColumns.included)
+            .isDefined
+        }(plan, index)
+      }
+
+    Map(indexes.head._1 -> candidateIndexes)
+  }
+}
+
+object HPFilterRankFilter extends IndexRankFilter {
+  override def apply(plan: LogicalPlan, indexes: PlanToIndexesMap): PlanToSelectedIndexMap = {
+    if (indexes.isEmpty || indexes.size != 1 || indexes.head._2.isEmpty) {
+      Map.empty
+    } else {
+      val selected = FilterIndexRanker.rank(spark, plan, indexes.head._2).get
+      setRankReasonTag(indexes.head._1, indexes.head._2, selected)
+      Map(indexes.head._1 -> selected)
+    }
+  }
+}
+
+/**
+ * HPFilterIndexRule looks for opportunities in a logical plan to replace
+ * a relation with an available hash partitioned index according to columns in
+ * filter predicate.
+ */
+object HPFilterIndexRule extends HyperspaceRule {
+  override val filtersOnQueryPlan
+    : Seq[QueryPlanIndexFilter] = FilterPlanNodeFilter :: HPFilterColumnFilter :: Nil
+
+  override val indexRanker: IndexRankFilter = HPFilterRankFilter
+
+  override def applyIndex(plan: LogicalPlan, indexes: PlanToSelectedIndexMap): LogicalPlan = {
+    if (indexes.isEmpty || (indexes.size != 1)) {
+      return plan
+    }
+
+    // As FilterIndexRule is not intended to support bucketed scan, we set
+    // useBucketUnionForAppended as false. If it's true, Hybrid Scan can cause
+    // unnecessary shuffle for appended data to apply BucketUnion for merging data.
+    RuleUtils.transformPlanToUseIndex(
+      spark,
+      indexes.head._2,
+      plan,
+      useBucketSpec = HyperspaceConf.useBucketSpecForFilterRule(spark),
+      useBucketUnionForAppended = false)
+  }
+
+  override def score(plan: LogicalPlan, indexes: PlanToSelectedIndexMap): Int = {
+    if (indexes.isEmpty || (indexes.size != 1)) {
+      return 0
+    }
+
+    val candidateIndex = indexes.head._2
+    // Filter index rule
+    val relation = RuleUtils.getRelation(spark, plan).get
+    val commonBytes = candidateIndex
+      .getTagValue(relation.plan, IndexLogEntryTags.COMMON_SOURCE_SIZE_IN_BYTES)
+      .getOrElse {
+        relation.allFileInfos.foldLeft(0L) { (res, f) =>
+          if (candidateIndex.sourceFileInfoSet.contains(f)) {
+            (res + f.size) // count, total bytes
+          } else {
+            res
+          }
+        }
+      }
+
+    // TODO enhance scoring
+    (50 * (commonBytes * 1.0f / relation.allFileSizeInBytes)).round
```

Comment:
1. I'll set left child `70` + right child `70` for JoinIndexRule so that it has a higher priority than FilterIndexRule
2. `1.0f` is just for float calculation and one coding habit from my C++ impl. I will try to use toFloat.